llama_stack/distribution/distribution.py

```diff
@@ -59,11 +70,65 @@ def providable_apis() -> List[Api]:
     return [api for api in Api if api not in routing_table_apis and api != Api.inspect and api != Api.providers]
 
 
-def get_provider_registry() -> Dict[Api, Dict[str, ProviderSpec]]:
+def get_provider_registry(config: Optional[StackRunConfig] = None) -> Dict[Api, Dict[str, ProviderSpec]]:
     ret = {}
     for api in providable_apis():
         name = api.name.lower()
+        logger.debug(f"Importing module {name}")
         module = importlib.import_module(f"llama_stack.providers.registry.{name}")
         ret[api] = {a.provider_type: a for a in module.available_providers()}
 
+    # Load external provider specs from YAML files if specified
+    # Walk through the external providers directory and load the provider specs
+    # for each API. A provider spec is a YAML file that contains the provider type,
+    # the adapter, and the API dependencies. Here is a tree of the directory structure:
+    #
+    # external_providers/
+    #   inference/
+    #     ollama.yaml
+    #     vllm.yaml
+    #   shields/
+    #     llama-guard.yaml
+    #     openai.yaml
+    #   vector_io/
+    #     qdrant.yaml
+    #     pinecone.yaml
+    #     weaviate.yaml
+    #   models/
+    if config and config.external_providers_dir:
+        external_providers_dir = os.path.abspath(config.external_providers_dir)
+        logger.info(f"Loading external providers from {external_providers_dir}")
+
+        for api in providable_apis():
+            api_dir = os.path.join(external_providers_dir, api.name.lower())
+            if not os.path.exists(api_dir):
+                logger.debug(f"No external provider directory found for {api.name}")
+                continue
+
+            # Look for provider spec files in the API directory
+            for filename in os.listdir(api_dir):
+                if not filename.endswith(".yaml"):
+                    logger.debug(f"Skipping non-YAML file {filename}")
+                    continue
+
+                provider_type = os.path.splitext(filename)[0]
+                spec_path = os.path.abspath(os.path.join(api_dir, filename))
+                logger.info(f"Loading provider(s) spec from {spec_path}")
+
+                try:
+                    with open(spec_path, "r") as f:
+                        spec_data = yaml.safe_load(f)
+
+                    # Convert YAML data to ProviderSpec
+                    adapter = AdapterSpec(**spec_data["adapter"])
+                    spec = remote_provider_spec(
+                        api=api,
+                        adapter=adapter,
+                        api_dependencies=[Api(dep) for dep in spec_data.get("api_dependencies", [])],
+                    )
+                    provider_type = f"remote::{adapter.adapter_type}"
```

Comment:
this being hardcoded as `remote` seems wrong (in some cases). Should this be part of dir structure similar to the in-tree? Or have a new root type, e.g. `external::`?

Reply:
I originally used a root type to represent the provider type but I wasn't sure if it was needed. Does it make sense to have an inline out-of-tree provider?

Reply:
I think so. How are they different in terms of whether someone would like to build a custom provider without contributing it upstream in-tree?

Reply:
You're right, let me change to reflect that.

Reply:
Updated!